text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,325 @@
+│                │                │   Build.shellScript   │
+│  CdkPipeline   │ Source.gitHub  │    (Vpc, CodeBuild    │
+│                │(SecretsManager)│     testing, IAM)     │
+│                │                │                       │
+├────────────────┴────────────────┴───────────────────────┤
+│                                                         │
+│            Workflow core + CDK app knowledge            │
+│   (steps, dependencies, translate CDK app into steps)   │
+│                                                         │
+├─────────────────────────────────────────────────────────┤
+│                                                         │
+│            Render to CodePipeline/CodeBuild             │
+│                                                         │
+└─────────────────────────────────────────────────────────┘
+```
+
+The **middle** layer has facilities to build and manipulate an abstract workflow, which features concepts like *steps*,
+*dependencies*, *artifacts*, and knows how to translate a CDK app into a sequence of backend-agnostic steps.
+
+The **bottom** layer renders the generic workflow to a specific CI/CD runner.  In the CDK Pipelines case, a CodePipeline
+with a set of CodeBuild projects and CloudFormation Actions.
+
+The **top** layer is the one the user interacts with. It is specialized to the backend that will be used to render out
+the final pipeline, and so can deal in concepts that are familiar to the user and offer all the backend-specific customizations
+users might want (Most commonly: source customizations such as credentials coming from AWS SecretsManager, and CodeBuild customizations
+like VPC bindings, specific additional IAM permissions, enabling CodeBuild test reporting, customizing Log Groups, etc).
+
+### How will we port to another backend?
+
+What is the work to implement different backends for CDK deployments (like, for example, GitHub Actions)?
+
+* Obviously, we need to replace layer 3
+* We will also need to replace layer 1. This is also pretty obvious: for other backends, the set of
+  supported Sources will be different, how we specify credentials will be different, the customizations offered
+  will be different. The most natural API will be one that maps closely to the platform we're targeting.
+  The `CdkPipeline` class (which currently takes `codepipeline.IAction`s, and `IPipeline`s, and will continue to do so
+  for backwards compatibility) will also be reimplemented. For example, to a class like `CdkAppWorkflow`.
+
+We will gain efficiency from the fact that the reimplementation will not contain business logic: it will
+mostly be mapping types and properties between the different levels. All the knowledge about the structure
+and order of deploying CDK apps will be located in layer 2, and is reused.
+
+### Why will we stay backwards compatible?
+
+Many people have already invested a lot of effort into writing code against the CDK Pipelines
```

Comment:
As mentioned above. This is nice-to-have but not a goal in itself in my mind, and we can always provide a shim layer for backwards compatibility to allow dev-previewers to migrate.

Let's make sure this is not influencing our design.